sycl/source/detail/device_image_impl.hpp

```diff
@@ -0,0 +1,127 @@
+//==------- device_image_impl.hpp - SYCL device_image_impl -----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/context.hpp>
+#include <CL/sycl/detail/common.hpp>
+#include <CL/sycl/detail/pi.h>
+#include <CL/sycl/device.hpp>
+#include <CL/sycl/kernel_bundle.hpp>
+#include <detail/device_impl.hpp>
+#include <detail/kernel_id_impl.hpp>
+#include <detail/program_manager/program_manager.hpp>
+
+#include <algorithm>
+#include <cassert>
+#include <cstring>
+#include <memory>
+#include <vector>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+class device_image_impl {
+public:
+  device_image_impl(RTDeviceBinaryImage *BinImage, context Context,
+                    std::vector<device> Devices, bundle_state State)
```

Comment:
Could you please tell if this is PR blocking comment?

Reply:
I'd say yes, unless you have specific reasons for passing by value and using std::move then. Or at least this should be made consistent with kernel_bundle_impl below, otherwise this non-uniformity seems confusing.

BTW, according to https://stackoverflow.com/questions/51705967/advantages-of-pass-by-value-and-stdmove-over-pass-by-reference/51706522 (2nd answer), if you really want to optimize (to me, does not seem worth it here), perfect forwarding seems a better solution.

Reply:
The reason for such a syntax is to avoid copies if an object is constructed like this:

`device_image_impl NewObj(BinImage, std::move(Context), std::move(Devices), State);`

It seems the answer from your comment explains that.

Reply:
Done.